### PR TITLE
Add configuration guide

### PR DESCRIPTION
**Pull request by mamhoff**
This commit adds a very high-level guide to some of the ways
it's possible to customize Solidus:

* Configuration options
* Overriding templates
* Changing assets
* Installing extensions
* Custom service classes
* Monkey-Patching Solidus
* Run a private fork

While we like some of these methods better than others, I think it's
important for beginners to understand their options.

### CODE REVIEW

File: guides/configuration.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# Configuration
+
+When you first setup a Rails app with Solidus installed, you will want to customize its looks and behaviour to fit your company's needs. Solidus can be customized in six ways:
+
+* Configuration options
```
Comment (tvdeyen) [inline guides/configuration.md]: Is it possible to link these?

File: guides/configuration.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+* Overriding templates
+* Changing assets
+* Installing extensions
+* Custom service classes
+* Monkey-Patching Solidus
+* Run a private fork
+
+You will most probably use at least methods 1-4, and most shops we know also use some combination of the last three.
+
+As Solidus is an Engine, much of what the [Spree Guide on using Engines](http://guides.rubyonrails.org/engines.html) explains applies directly to Solidus.
```
Comment (tvdeyen) [inline guides/configuration.md]: > ~~Spree~~ Guide on using Engines

Rails Guide on using Engines

File: guides/configuration.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+## Configuration options
+
+Since Solidus is a [Rails Engine](http://guides.rubyonrails.org/engines.html), much of its behaviour can be customized through the Solidus initializer. After you have run the `rails g spree:install` generator, this initializer will live in `config/initializers/spree.rb`. Open this file, and have a look at the first block (comments left out):
```
Comment (tvdeyen) [inline guides/configuration.md]: Maybe explaining the `spree` vs. `solidus` incidence in the introductory section would be helpful to understand the differing file names especially for the unaware.

File: guides/configuration.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+For example, if you want to change the way the product list is rendered, take the raw version of [that template] (https://github.com/solidusio/solidus/blob/master/frontend/app/views/spree/products/index.html.erb) and copy it to `app/views/spree/products/index.html.erb`. It is very important that the path and name of this file are exactly the same as in the Gem! Template files from your app will always win over template files from an imported engine.
+
+Of course, it the template you wish to modify references other template files, you can override those, too. For example, the product list references a shared partial `app/views/spree/shared/_products.html.erb` - you can find it [here](https://github.com/solidusio/solidus/blob/master/frontend/app/views/spree/shared/_products.html.erb) and also override it in your host app.
+
+When you override a view from a Gem, there is the possibility of difficulties when upgrading, because Controller code within the Gem might have changed, such as for example the name of an instance variable. Be careful when upgrading, and always read `CHANGELOG.md`.
```
Comment (tvdeyen) [inline guides/configuration.md]: Maybe mark this as side note or tip

File: guides/configuration.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+### In the backend
+
+We do not recommend overriding the admin backend in the same way, since some of the extensions you might be running (see below) depend on them having the HTML structure present in the version of Solidus they were built for. If you need to add custom functionality to Solidus' backend, we recommend using the `Deface` gem. Find it [here](https://github.com/spree/deface). Overriding backend views with Deface is tricky - be sure to have good tests!
```
Comment (tvdeyen) [inline guides/configuration.md]: Do we recommend the usage of Deface?

File: guides/configuration.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+## Changing assets
+
+If you need to change the look and feel of your app without changing the HTML structure, you can customize the CSS, image and JavaScript assets by amending the files in `vendor/assets/*/*/all.js`. They mirror the structure of the `app/assets` folder. Have a look at the [Spree Guide](http://guides.spreecommerce.org/developer/asset.html) for overriding assets, it applies to Solidus as-is.
```
Comment (tvdeyen) [inline guides/configuration.md]: I think it's kinda dangerous to link to the spree guides. We can't rely on them being online forever.

File: guides/configuration.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+## Installing extensions
+
+The Solidus ecosystem is rich in extensions that modify Solidus' behaviour. The extensions are distributed as Gems, so it's often as easy as adding the extension to your Gemfile and running an install generator. Sometimes, however, the extension needs more work, such as adding custom configuration in its own initializer. Please refer to the individual extension's Readme file.
+
+We maintain an [extensions compatibility matrix](http://extensions.solidus.io/). This is a good starting point for finding out which extensions are available and whether they work with your chosen Solidus version.
+
+Every extension adds another dependency to your app. Choose wisely.
```
Comment (tvdeyen) [inline guides/configuration.md]: Good advice, but unnecessary IMO

File: guides/configuration.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+## Custom service classes
+
+The Solidus configuration object from above has some getters and setters for classes as extension points. After evaluating whether you need to go down this route, take some time to study the standard implementation for the class you want to modify, and subclass or re-implement it. The service class extension points can be found [here](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/app_configuration.rb#L277-L383).
+
+These extension points can alter Solidus' behaviour quite drastically. If you make mistakes when implementing them, your store will break. Good luck!
```
Comment (tvdeyen) [inline guides/configuration.md]: Nah, this is too harsh. I think we can just remove this line

File: guides/configuration.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+## Monkey-Patching Solidus
+
+Solidus is setup so that any file in the `app` directory with the suffix `_decorator.rb` will be auto-loaded like any model or controller files.
+
+For example, if you want to add a method to the `Spree::Order` model, you would have a file called `app/models/spree/order_decorator.rb` with the following contents:
+
+```
+module Spree
```
Comment (tvdeyen) [inline guides/configuration.md]: It's not necessary to use the `Spree` `module` namespace and can even lead to strange issues. 

Maybe just use none or `module YourStore`

File: guides/configuration.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  module OrderDecorator
+    def total
+      "One Million Dollars"
+    end
+  end
+end
+
+Spree::Order.prepend Spree::OrderDecorator
```
Comment (tvdeyen) [inline guides/configuration.md]: Great usage of `Module.prepend`

"No more `class_eval`" 👌

File: guides/configuration.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+```
+
+This creates a new module called `Spree::OrderDecorator` and will insert its methods so early in the method lookup chain that the decorator's methods overwrite the original order class's methods.
```
Comment (tvdeyen) [inline guides/configuration.md]: I think you don't need to explain `Module.prepend` here. Maybe link to a good article?

File: guides/configuration.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+From now on, every order, when asked for its total, will return the string "One Million Dollars". This particular decorator would, by the way, definitely break a large part of Solidus' functionality as `Spree::Order#total` is expected to return a `BigDecimal`, not a `String`.
+
+If you do this kind of thing, be very careful and test your new functionality very well. If you do upgrades, read extra carefully about changes in Solidus' core classes.
```
Comment (tvdeyen) [inline guides/configuration.md]: Also mark this as "tip" or "side note"?

File: guides/configuration.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+Spree::Order.prepend MyStore::OrderDecorator
+```
+
+This creates a new module called `MyStore::OrderDecorator` and will insert its methods so early in the method lookup chain for method calls on `Spree::Order` objects that the decorator's methods override the original methods.
+
+From now on, every order, when asked for its total, will return the an inflated total by 10 Dollars (or whatever your currency is).
+
+If you do this kind of thing, be very careful and test your new functionality very well. If you do upgrades, read extra carefully about changes in Solidus' core classes.
+
+## Run a private fork (and submit us a Pull Request!)
```
Comment (tvdeyen) [inline guides/configuration.md]: Could you please remove the call for pull requests? This is a little bit too "loud" ;)

Comment (cbrunsdon) [inline guides/configuration.md]: This breaks the link at the top, as well.

**Pull request merged**